Source/JavaScriptCore/runtime/JSONObject.cpp

```diff
@@ -627,6 +637,517 @@ bool Stringifier::Holder::appendNextProperty(Stringifier& stringifier, StringBui
+            if (UNLIKELY(scope.exception())) {
+                scope.clearException();
+                recordFailure("getOwnPropertyNames exception"_s);
+                return;
+            }
+            for (auto& name : names) {
+                if (!name.string().is8Bit()) {
+                    recordFailure("16-bit property name"_s);
+                    return;
+                }
+                bool needComma = m_buffer[m_length - 1] != '{';
+                unsigned nameLength = name.length();
+                if (remainingCapacity() < needComma + 1 + nameLength + 2) {
+                    recordBufferFull();
+                    return;
+                }
+                if (needComma)
+                    m_buffer[m_length++] = ',';
+                m_buffer[m_length] = '"';
+                auto* characters = name.string().characters8();
+                for (unsigned i = 0; i < nameLength; ++i) {
+                    auto character = characters[i];
+                    if (mayNeedJSONEscaping(character)) {
+                        recordFailure("property name character needs escaping"_s);
+                        return;
+                    }
+                    m_buffer[m_length + 1 + i] = character;
+                }
+                m_buffer[m_length + 1 + nameLength] = '"';
+                m_buffer[m_length + 1 + nameLength + 1] = ':';
+                m_length += 1 + nameLength + 2;
+                scope.assertNoException();
+                auto value = object.get(&m_globalObject, name);
+                if (UNLIKELY(scope.exception())) {
+                    scope.clearException();
+                    recordFailure("get exception"_s);
+                    return;
+                }
+                append(value);
+                if (haveFailure())
+                    return;
+            }
+#endif
+        }
+        if (!remainingCapacity()) {
+            recordBufferFull();
+            return;
+        }
+        m_buffer[m_length++] = '}';
+        return;
+    }
+
+    case ArrayType: {
+        auto& array = *asArray(&cell);
+        if (array.isCallable()) {
+            recordFailure("callable array"_s);
+            return;
+        }
+        if (!m_checkedArrayPrototype) {
+            if (mayHaveToJSON(*m_globalObject.arrayPrototype())) {
+                recordFailure("array prototype may have toJSON"_s);
+                return;
+            }
+            m_checkedArrayPrototype = true;
+        }
+        if (!remainingCapacity()) {
+            recordBufferFull();
+            return;
+        }
+        array.structure()->forEachProperty(m_vm, [&](const PropertyTableEntry& entry) -> bool {
+            if (entry.key() == m_vm.propertyNames->toJSON) {
+                recordFailure("array has toJSON"_s);
+                return false;
+            }
+            return true;
+        });
```

Comment:
Here is a tip. Most of arrays do not have their own custom property. So, let's add a fast path,

```
if (m_globalObject.isOriginalArrayStructure(array.structure())) {
    // Then, there is no toJSON property. This is original array structure, which means, no custom properties are added.
} else {
    // Check toJSON in the own properties.
}
```

Reply:
Great idea. Will do.

Reply:
Done.